api/handler.go

```diff
@@ -42,48 +67,104 @@ type apiHandler struct {
 func NewHandler(serverCtx context.Context, opts APIHandlerOptions, healthcore *health.Core, views *views.Client) http.Handler {
 	handler := &apiHandler{opts, serverCtx, healthcore, views}
 
-	router := httprouter.New()
-	router.HandlerFunc("GET", "/_healthz", handler.healthcheck)
+	router := chi.NewRouter()
+
+	// don't use middlewares for the system routes
+	router.Get("/_healthz", handler.healthcheck)
 	if opts.Prometheus {
-		router.Handler("GET", "/metrics", promhttp.Handler())
+		router.Method("GET", "/metrics", promhttp.Handler())
 	}
-	addStreamHealthHandlers(router, handler)
-	addViewershipHandlers(router, handler)
 
-	globalMiddlewares := []middleware{handler.cors()}
-	return prepareHandler("", false, router, globalMiddlewares...)
+	router.Route(opts.APIRoot, func(router chi.Router) {
+		router.Use(chimiddleware.Logger)
+		router.Use(chimiddleware.NewCompressor(5, "application/json").Handler)
+		router.Use(handler.cors())
+
+		router.Mount(`/stream/{`+streamIDParam+`}`, handler.streamHealthHandler())
+		router.Mount("/views", handler.viewershipHandler())
+	})
+
+	return router
 }
 
-func addStreamHealthHandlers(router *httprouter.Router, handler *apiHandler) {
-	healthcore, opts := handler.core, handler.opts
-	middlewares := []middleware{
+// {streamId} variable must be set in the request context
+func (h *apiHandler) streamHealthHandler() chi.Router {
+	healthcore, opts := h.core, h.opts
+
+	router := chi.NewRouter()
+	router.Use(
 		streamStatus(healthcore),
-		regionProxy(opts.RegionalHostFormat, opts.OwnRegion),
-	}
+		regionProxy(opts.RegionalHostFormat, opts.OwnRegion))
 	if opts.AuthURL != "" {
-		middlewares = append(middlewares, authorization(opts.AuthURL))
-	}
-	addApiHandler := func(apiPath, name string, handler http.HandlerFunc) {
-		fullPath := path.Join(opts.APIRoot, "/stream/:"+streamIDParam, apiPath)
-		fullHandler := prepareHandlerFunc(name, opts.Prometheus, handler, middlewares...)
-		router.Handler("GET", fullPath, fullHandler)
+		router.Use(authorization(opts.AuthURL))
 	}
-	addApiHandler("/health", "get_stream_health", handler.getStreamHealth)
-	addApiHandler("/events", "stream_health_events", handler.subscribeEvents)
+
+	h.withMetrics(router, "get_stream_health").
+		MethodFunc("GET", "/health", h.getStreamHealth)
+	h.withMetrics(router, "stream_health_events").
+		MethodFunc("GET", "/events", h.subscribeEvents)
+
+	return router
 }
 
-func addViewershipHandlers(router *httprouter.Router, handler *apiHandler) {
-	opts := handler.opts
-	middlewares := []middleware{}
+func (h *apiHandler) viewershipHandler() chi.Router {
+	opts := h.opts
+
+	router := chi.NewRouter()
 	if opts.AuthURL != "" {
-		middlewares = append(middlewares, authorization(opts.AuthURL))
+		router.Use(authorization(opts.AuthURL))
```

Comment:
Get total views shouldn't have the auth middleware

Reply:
Discussed offline and decided we should still keep it on the /total views API. Let's have the same authorization for this API as we have for the playback info endpoint.